src/Microsoft.Android.Sdk.ILLink/SetupStep.cs

```diff
@@ -26,36 +27,35 @@ class SetupStep : BaseStep
 			}
 		}
 
-		static MethodInfo getReferencedAssembliesMethod = typeof (LinkContext).GetMethod ("GetReferencedAssemblies", BindingFlags.Public | BindingFlags.Instance);
+		List<IMarkHandler> MarkHandlers {
+			get {
+				if (_markHandlers == null) {
+					var pipeline = typeof (LinkContext).GetProperty ("Pipeline").GetGetMethod ().Invoke (Context, null);
+					_markHandlers = (List<IMarkHandler>) pipeline.GetType ().GetProperty ("MarkHandlers").GetValue (pipeline);
+				}
+				return _markHandlers;
+			}
+		}
```

Comment:
@sbomer I think this should be avoided and use a supported registration mechanism

Reply:
I will look into updating it to use multiple custom step arguments instead of SetupStep

Reply:
This is possible, but messy because separate custom steps don't have a great way to share state. Either:
1. Steps with shared state need to be combined into one (which imposes ordering constraints and isn't great for code organization), or
2. Shared state needs to be static, which hits this bug: https://github.com/mono/linker/issues/1314.

It might be better to leave as-is, and make some of the Pipeline methods part of the public API for custom steps.
If we do want to do 1 or 2, I'd rather do it in a separate PR.

Reply:
I'd like to see the reflection code removed as it's very fragile.

Reply:
The change which adds public API for resolve cache landed and the API is available on LinkerContext.

@sbomer could you update the PR to use it instead of the one from xamarin-android?

Reply:
Done, PTAL. The LinkContextExtensions logic is copied and adjusted from https://github.com/xamarin/java.interop/tree/main/src/Java.Interop.Tools.Cecil/Java.Interop.Tools.Cecil. Another option if we want to avoid duplication would be to introduce a dependency on illink in java.interop - please let me know what would be best.

Reply:
It looks like this project links in a few files from `Java.Interop.Tools.Cecil`, can you add a new file here:

https://github.com/xamarin/xamarin-android/blob/af7f7f5da475c5cb8d2d725d77082ba7915bbc7e/src/Microsoft.Android.Sdk.ILLink/Microsoft.Android.Sdk.ILLink.csproj#L43-L46

We probably should make this a regular project reference, but you don't have to do that here -- we can do it later.

Reply:
I'm not sure I understand - are you saying you'd prefer the LinkContextExtensions to live in the Java.Interop repo? The way I have it now, they live in the Microsoft.Android.Sdk.ILLink project so they are already included as compile items.

Reply:
I guess I'm not following why we need two versions of the extension methods.

Could we create a `TypeDefinitionCache` in .NET 6, or did that cause a problem? We could create one at the step level, if we don't want to share it between steps.

Reply:
Ah, sorry. The shared `TypeDefinitionCache` doesn't work well with .NET 6, and the new extension methods use a cache that's built-in to the linker instead.

More detail:
The `TypeDefinitionCache` shared across steps worked fine before this change, but only because the steps were injected into the pipeline using reflection (with one `--custom-step` to bootstrap and set up the others). The "supported" API only allows adding steps using the `--custom-step` argument, which doesn't play nicely with shared state (https://github.com/mono/linker/issues/1314).

A step-level cache would still work, but could reduce the advantages of caching. @marek-safar instead added a shared cache to the linker itself in https://github.com/mono/linker/pull/1979, and suggested we use that one here, so I added those extension methods to use the built-in shared cache. This doesn't solve the shared state problem in general, but this was the only shared state being used in xamarin-android.

Other alternatives would be:
- Leave it as-is (continue using a shared `TypeDefinitionCache` instead of the built-in one, using reflection to inject the steps)
- Introduce a supported way to share state between steps (see https://github.com/mono/linker/issues/1953) and use a shared `TypeDefinitionCache`
- Avoid using reflection but forego the shared cache - only have one `TypeDefinitionCache` per step.

@marek-safar wanted the reflection to go away, and wasn't sure that the `TypeDefinitionCache` was a compelling reason to extend the custom step API since it would be useful as a shared cache. Now that the shared cache is available I don't see much of an advantage to having a cache per step. Let me know what you think!

Reply:
I don't think custom steps should have their own type/method/fields resolving caching. It's redundant and it also usually does not work well with the linker settings which control the resolving process